### PR TITLE
fix(Scripts/HyjalSummit): Adjust Doomfire Spirit movement logic.

### PR DESCRIPTION
**Pull request by heyitsbench**
## Changes Proposed:

This PR proposes changes to:
-  [X] Scripts (creature scripts).

## Issues Addressed:

- Closes unreported issue.

## SOURCE:

The changes have been validated through:
- [X] Sniffs.
- [X] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.).

## Tests Performed:

This PR:
- [X] Has not been tested.

## How to Test the Changes:

- [X] This pull request requires further testing.
1. Engage Archimonde.
2. Wait for Doomfire.
3. Observe its movement.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by heyitsbench** [comment]
Doomfire sucks by the way.

File: src/server/scripts/Kalimdor/CavernsOfTime/BattleForMountHyjal/boss_archimonde.cpp [hunk 1]
```diff
@@ -145,32 +145,29 @@ struct npc_doomfire_spirit : public ScriptedAI
...
+            float dist = 8.f;
 
-        fAngle = urand(0, M_PI * 2);
-    }
+            while (dist >= 0)
+            {
+                pos = me->WorldObject::GetFirstCollisionPosition(dist, nextOrientation);
+                if (fabsf(dist - me->GetExactDist2d(pos)) < 0.001) // Account for small deviation
+                    break;
+                dist -= 2.0; // Distance drops by two units with each unsuccessful attempt
```
Comment (Kitzunu) [inline src/server/scripts/Kalimdor/CavernsOfTime/BattleForMountHyjal/boss_archimonde.cpp:163]:
```suggestion
                dist -= 2.0f; // Distance drops by two units with each unsuccessful attempt
```

**Comment by Gultask** [comment]
does this conflict or make https://github.com/azerothcore/azerothcore-wotlk/pull/19467 obsolete in any way? :pepeinspect:

**Comment by heyitsbench** [comment]
> does this conflict with or make 19467 obsolete in any way?

Not that I can discern, no.

**Comment by Tereneckla** [comment]
> does this conflict with or make #19467 obsolete in any way? :pepeinspect:

That one is the starting direction and this one is behavior around obstacles

**Comment by heyitsbench** [comment]
So the way Doomfire seems to work is like this (just to summarize a bunch of things learned yesterday):
Doomfire and Doomfire Spirit both spawn at a given time, the Spirit might face Archimonde with its initial spawn.
~0.01s later, the Spirit does a 1y teleport away and has its orientation flipped 180 degrees.
~0.01s later, the Spirit then does an 8y teleport in the same style as its normal teleports.
~1.6s later, the Doomfire takes on its aura to create Doomfire area aura patches on the ground and begins following the Doomfire Spirit. At the same time, the Spirit teleports again in its usual fashion.

The way the teleport works is the Spirit seems to pick a random direction, either straight, 45 degrees left, or 45 degrees right of its current orientation. The selection of directions does not seem to be weighted based on the sample size we have. Once a direction is selected, it does a check in that direction with a distance of 8y if the coordinates are clear of obstruction. If it's successful, it teleports. If not, it checks again at 6y, then 4y, then 2y. If all three fail, then the Doomfire will still teleport, but nothing about its position nor orientation will change. Its subsequent teleport(s) will still pick direction based off the unchanged orientation.

After this PR, the Doomfire does not wait to "wake up" and take on its aura, nor do the first two Spirit teleports occur, only at 1.6s forward. The Doomfire issue is, in my opinion, [a symptom of larger missing behavior](https://github.com/azerothcore/azerothcore-wotlk/issues/19062), but the Spirit issue is likely easy to solve with some scheduling stuff.

**Comment by heyitsbench** [comment]
After the recent commits (thanks to Tereneckla), the first two Spirit teleports now are occurring. A misuse of `WorldObject::GetFirstCollisionPosition` was also corrected. The Doomfire itself still does not wait to start acting, but looking roughly at the fire, it does seem to exhibit closer behavior to what I've seen in videos and on Classic.

**Review by Nyeriah — APPROVED** [review]

**Pull request merged**